libcaf_core/src/telemetry/importer/process.cpp

```diff
@@ -0,0 +1,229 @@
+
+#ifdef CAF_HAS_PROCESS_METRICS
+
+namespace {
+
+struct sys_stats {
+  int64_t rss;
+  int64_t vms;
+  double cpu_time;
+};
+
+sys_stats read_sys_stats();
+
+static constexpr bool platform_supported_v = true;
+
+template <class CpuPtr, class RssPtr, class VmsPtr>
+void sys_stats_init(caf::telemetry::metric_registry& reg, RssPtr& rss,
+                    VmsPtr& vms, CpuPtr& cpu) {
+  rss = reg.gauge_singleton("process", "resident_memory",
+                            "Resident memory size.", "bytes");
+  vms = reg.gauge_singleton("process", "virtual_memory", "Virtual memory size.",
+                            "bytes");
+  cpu = reg.gauge_singleton<double>("process", "cpu",
+                                    "Total user and system CPU time spent.",
+                                    "seconds", true);
+}
+
+void update_impl(caf::telemetry::int_gauge* rss_gauge,
+                 caf::telemetry::int_gauge* vms_gauge,
+                 caf::telemetry::dbl_gauge* cpu_gauge) {
+  auto [rss, vmsize, cpu] = read_sys_stats();
+  rss_gauge->value(rss);
+  vms_gauge->value(vmsize);
+  cpu_gauge->value(cpu);
+}
+
+} // namespace
+
+#else // CAF_HAS_PROCESS_METRICS
+
+namespace {
+
+static constexpr bool platform_supported_v = false;
+
+template <class... Ts>
+void sys_stats_init(Ts&&...) {
+  // nop
+}
+
+template <class... Ts>
+void update_impl(Ts&&...) {
+  // nop
+}
+
+} // namespace
+
+#endif // CAF_HAS_PROCESS_METRICS
+
+// -- macOS-specific scraping logic --------------------------------------------
+
+#ifdef CAF_MACOS
+
+#  include <mach/mach.h>
+#  include <mach/task.h>
+#  include <sys/resource.h>
+
+namespace {
+
+sys_stats read_sys_stats() {
+  sys_stats result{0, 0, 0};
+  // Fetch memory usage.
+  {
+    mach_task_basic_info info;
+    mach_msg_type_number_t count = MACH_TASK_BASIC_INFO_COUNT;
+    if (task_info(mach_task_self(), MACH_TASK_BASIC_INFO, (task_info_t) &info,
+                  &count)
+        == KERN_SUCCESS) {
+      result.rss = info.resident_size;
+      result.vms = info.virtual_size;
+    }
+  }
+  // Fetch CPU time.
+  {
+    task_thread_times_info info;
+    mach_msg_type_number_t count = TASK_THREAD_TIMES_INFO_COUNT;
+    if (task_info(mach_task_self(), TASK_THREAD_TIMES_INFO, (task_info_t) &info,
+                  &count)
+        == KERN_SUCCESS) {
+      // Round to milliseconds.
+      result.cpu_time += info.user_time.seconds;
+      result.cpu_time += ceil(info.user_time.microseconds / 1000.0) / 1000.0;
+      result.cpu_time += info.system_time.seconds;
+      result.cpu_time += ceil(info.system_time.microseconds / 1000.0) / 1000.0;
+    }
+  }
+  return result;
+}
+
+} // namespace
+
+#endif // CAF_MACOS
+
+// -- Linux-specific scraping logic --------------------------------------------
+
+#ifdef CAF_LINUX
+
+#  include <cstdio>
+#  include <unistd.h>
+
+namespace {
+
+std::atomic<long> global_ticks_per_second;
+
+std::atomic<long> global_page_size;
```

Comment:
What are these here for? They are not used anywhere ... at least that's what my grep says.

Reply:
They are used in `TRY_LOAD` to cache the result of syscalls.

Reply:
Ah, thanks.